### ISSUE TITLE
🎯 Multi-Level Predictions: RUL + Component-Level

### ISSUE THREAD

**Issue by Shukik85** [Shukik85/hydraulic-diagnostic-saas#116]
## 🎯 Objective

Extend UniversalTemporalGNN with multi-level predictions:
- Graph-level: Add RUL (Remaining Useful Life) prediction
- Component-level: Per-node health & anomaly predictions

**Epic:** #92 (GNN Service - Production-Ready Implementation)
**Phase:** Phase 2 - Training & Integration (Week 2)
**Priority:** High
**Estimate:** 3-4 hours

---

### 1. RUL Prediction Head (1h)

- [ ] Add QuantileRULLoss class
  - Asymmetric penalties (underestimation > overestimation)
  - Multiple quantiles (0.1, 0.5, 0.9)
  - Log-normalization support

- [ ] Add RUL head to UniversalTemporalGNN
  - Sequential(Linear(lstm_hidden, 128) → ReLU → Dropout → Linear(128, 1) → Softplus)
  - Output: [B, 1] hours until failure

- [ ] Update schemas
  - Add `rul_hours` to PredictionResponse
  - Update ModelInfo with RUL support

### 2. Component-Level Predictions (1.5h)

- [ ] Add component-level heads to UniversalTemporalGNN
  - `component_health`: Sequential(Linear(hidden, 64) → ReLU → Dropout → Linear(64, 1) → Sigmoid)
  - `component_anomaly`: Sequential(Linear(hidden, 128) → ReLU → Dropout → Linear(128, 9))
  - Output: health [N, 1], anomaly [N, 9]

- [ ] Update forward() method
  - Component predictions BEFORE pooling
  - Graph predictions AFTER LSTM
  - Return nested dict structure

- [ ] Update schemas
  - Add ComponentPrediction schema
  - Update PredictionResponse with component field

### 3. Training Integration (1h)

- [ ] Update HydraulicGNNModule
  - Support multi-level outputs
  - Per-level loss computation
  - Per-level metrics logging

- [ ] Update MultiTaskLoss
  - Add RUL loss integration
  - Support component-level losses
  - Proper loss aggregation

- [ ] Update training_step/validation_step
  - Handle nested outputs
  - Log all levels separately

### 4. Testing (30min)

- [ ] Unit tests for QuantileRULLoss
- [ ] Unit tests for component-level heads
- [ ] Integration test for multi-level forward pass
- [ ] Test multi-level loss computation

---

### QuantileRULLoss

```python
class QuantileRULLoss(nn.Module):
    """Quantile loss for RUL prediction.
    
    Asymmetric penalties:
    - Underestimation (predict too early) → higher penalty
    - Overestimation (predict too late) → lower penalty
    
    References:
        - https://openreview.net/forum?id=tzFjcVqmxw
    """
    
    def __init__(self, quantiles=[0.1, 0.5, 0.9]):
        super().__init__()
        self.quantiles = quantiles
    
    def forward(self, pred, target):
        losses = []
        for q in self.quantiles:
            error = target - pred
            loss = torch.max(
                q * error,
                (q - 1) * error
            )
            losses.append(loss)
        
        return torch.mean(torch.stack(losses))
```

### Updated Forward Method

```python
def forward(self, x, edge_index, edge_attr, batch):
    # GATv2 encoding
    h = self.initial_projection(x)
    for gat_layer in self.gat_layers:
        h = gat_layer(h, edge_index, edge_attr)
    
    # === Component-Level Predictions ===
    component_health = self.component_health(h)      # [N, 1]
    component_anomaly = self.component_anomaly(h)    # [N, 9]
    
    # === Graph-Level Predictions ===
    graph_repr = global_mean_pool(h, batch)          # [B, hidden]
    lstm_out = self.lstm(graph_repr.unsqueeze(1))    # [B, lstm_hidden]
    lstm_out = lstm_out.squeeze(1)
    
    # Graph-level heads
    graph_health = self.health_head(lstm_out)        # [B, 1]
    graph_degradation = self.degradation_head(lstm_out)  # [B, 1]
    graph_anomaly = self.anomaly_head(lstm_out)      # [B, 9]
    graph_rul = self.rul_head(lstm_out)              # [B, 1]
    
    return {
        'component': {
            'health': component_health,
            'anomaly': component_anomaly
        },
        'graph': {
            'health': graph_health,
            'degradation': graph_degradation,
            'anomaly': graph_anomaly,
            'rul': graph_rul
        }
    }
```

---

### Functionality
- ✅ RUL prediction works
- ✅ Component-level predictions work
- ✅ Multi-level loss computed correctly
- ✅ Nested output structure correct

### Code Quality
- ✅ Type hints (100%)
- ✅ Docstrings (100%)
- ✅ Unit tests pass
- ✅ Integration tests pass

### Performance
- ✅ No performance regression (<10% slower)
- ✅ Memory usage reasonable
- ✅ Training stable

---

## 📊 Deliverables

**Code:**
- `src/training/losses.py` - Add QuantileRULLoss
- `src/models/universal_temporal_gnn.py` - Multi-level heads
- `src/schemas/responses.py` - Update response schemas
- `src/training/lightning_module.py` - Multi-level training

**Tests:**
- `tests/unit/test_losses.py` - RUL loss tests
- `tests/unit/test_model.py` - Component-level tests
- `tests/integration/test_training.py` - Multi-level training test

---

## 🔗 Dependencies

**Requires:**
- Issue #115 (PyTorch Lightning) - 65% complete ✅
- Issue #94 (GNN Model) ✅

**Blocks:**
- Issue #118 (Dynamic Edge Features)
- Issue #119 (Edge-Level Predictions)

---

## 📝 Notes

**Key considerations:**
- RUL prediction requires log-normalization of targets
- Component-level predictions should NOT use LSTM (only GATv2)
- Multi-level outputs require careful batch handling
- Loss weights may need rebalancing with RUL added

**References:**
- Quantile RUL loss: https://openreview.net/forum?id=tzFjcVqmxw
- Multi-level GNN: https://arxiv.org/abs/2404.10324
- Component diagnostics: https://pmc.ncbi.nlm.nih.gov/articles/PMC11125296/

---

**Created:** 2025-11-29 17:30 MSK
**Target completion:** 2025-11-29 (today, 3-4h)
